### PR TITLE
Support recursive arrays in JSON when an item is an array

### PR DESCRIPTION
**Pull request by AL-377**
I have noticed that the origin outlines doesn't support recursive array type. To be clear, here is a example：

```python
from outlines.text.json_schema import build_regex_from_schema
from pydantic import BaseModel
from typing import List
import json 

class ItemModel(BaseModel):
    name:str

class ArrayModel(BaseModel):
    items:List[List[ItemModel]]

test_json = \
{
    "items": [[{"name":"item1"},{"name":"item2"}]], 
}


json_schema = ArrayModel.model_json_schema()
print(json.dumps(json_schema,indent=2))

regex = build_regex_from_schema(json.dumps(json_schema))
print(regex)

parsed = ArrayModel.model_validate_json(json.dumps(test_json))
print(parsed)

```

Run the code above, then we can get the json schema of  `ArrayModel` , but we will get a error from `build_regex_from_schema` when we meet the `List[List[]]` type property.

And now, I have modified the code in outlines and it works. You can check the `test_json_schema.py`

### CODE REVIEW

File: .DS_Store
Comment (brandonwillard) [inline .DS_Store:1]: Looks like this file was accidentally added.

File: test.py
Comment (brandonwillard) [inline test.py:1]: These tests need to be added to the existing test files in `pytest` format.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: .DS_Store
Comment (AL-377) [reply .DS_Store:1]: OK，I have delete that file

File: test.py
Comment (AL-377) [reply test.py:1]: Sorry I have missed the pre-commit step before, and now it's done

**Comment by AL-377** [comment]
It is strange that the pytest here seems to fail in`test_continuation.py`, I have passed the `test_continuation.py` in my local machine.

image

**Review by AL-377 — COMMENTED** [review]
The changes have done

**Review by brandonwillard — APPROVED** [review]

**Comment by brandonwillard** [comment]
@AL-377, this was a great fix/enhancement; much appreciated!  I've added some minor refactoring and it should be ready to merge once everything finishes in CI.

**Comment by AL-377** [comment]
> @AL-377, this was a great fix/enhancement; much appreciated! I've added some minor refactoring and it should be ready to merge once everything finishes in CI.

👍👍👍

**Comment by AL-377** [comment]
> Thank you for the contribution and sorry for the delay in reviewing. Will get to it soon.

Never mind！

**Pull request merged**